pkg/lib/oauth/handler/resolve.go

```diff
@@ -101,3 +102,64 @@ func validateRedirectURI(
 
 	return nil
 }
+
+func parseAuthzRedirectURI(
+	client *config.OAuthClientConfig,
+	httpProto httputil.HTTPProto,
+	httpOrigin httputil.HTTPOrigin,
+	domainWhitelist []string,
+	r protocol.AuthorizationRequest,
+) (*url.URL, protocol.ErrorResponse) {
+	if r.SettingsAction() == "" {
+		return parseRedirectURI(client, httpProto, httpOrigin, domainWhitelist, r)
+	}
+
+	redirectURI, err := url.Parse(r.RedirectURI())
+	if err != nil {
+		return nil, protocol.NewErrorResponse("invalid_request", "invalid redirect URI")
+	}
+
+	err = validatePostSettingsRedirectURI(client, httpProto, httpOrigin, domainWhitelist, redirectURI)
+	if err != nil {
+		return nil, protocol.NewErrorResponse("invalid_request", err.Error())
+	}
+
+	switch r.SettingsAction() {
+	case "change_password":
+		settingsRedirectURI := &url.URL{Path: "/settings/change_password"}
+		queryItems := settingsRedirectURI.Query()
+		queryItems.Set("redirect_uri", r.RedirectURI())
+		queryItems.Set("client_id", r.ClientID())
+
+		settingsRedirectURI.RawQuery = queryItems.Encode()
+		return settingsRedirectURI, nil
+	default:
+		return nil, protocol.NewErrorResponse("invalid_request", "invalid settings action")
+	}
+}
+
+func validatePostSettingsRedirectURI(
+	client *config.OAuthClientConfig,
+	httpProto httputil.HTTPProto,
+	httpOrigin httputil.HTTPOrigin,
+	domainWhitelist []string,
+	redirectURI *url.URL,
+) error {
+	redirectURIString := redirectURI.String()
+
+	matcher, err := originmatcher.New(client.SettingsRedirectURIOrigins)
+	if err != nil {
+		return err
+	}
+
+	if matcher.MatchOrigin(redirectURIString) {
+		return nil
+	}
+
+	err = validateRedirectURI(client, httpProto, httpOrigin, domainWhitelist, redirectURI)
+	if err != nil {
+		return err
+	}
+
+	return errors.New("redirect URI is not allowed")
```

Comment:
I think we should return nil here? If validateRedirectURI returns nil, that means the redirect URI is valid.